runtime/src/accounts_db.rs

```diff
@@ -2910,58 +2954,153 @@ impl AccountsDB {
                     "Flushing random slot: {}, num_remaining: {}",
                     *rand_slot, num_slots_remaining
                 );
-                self.flush_slot_cache(*rand_slot);
+                self.flush_slot_cache(*rand_slot, None::<&mut fn(&_, &_) -> bool>);
             }
         }
+    }
 
-        inc_new_counter_info!("flush_roots_elapsed", flush_roots_elapsed.as_us() as usize);
+    fn flush_rooted_accounts_cache(
+        &self,
+        mut max_flush_root: Option<Slot>,
+        should_clean: Option<(&mut usize, &mut usize)>,
+    ) -> (usize, usize) {
+        if should_clean.is_some() {
+            max_flush_root = self.max_clean_root(max_flush_root);
+        }
+
+        // If there is a long running scan going on, this could prevent any cleaning
+        // past `max_flush_root`.
+        let cached_roots: BTreeSet<Slot> = self.accounts_cache.clear_roots(max_flush_root);
+        let mut written_accounts = HashSet::new();
+
+        // If `should_clean` is None, then`should_flush_f` is also None, which will cause
+        // `flush_slot_cache` to flush all accounts to storage without cleaning any accounts.
+        let mut should_flush_f = should_clean.map(|(account_bytes_saved, num_accounts_saved)| {
+            move |pubkey: &Pubkey, account: &Account| {
+                // If a later root already wrote this account, no point
+                // in flushing it
+                let should_flush = !written_accounts.contains(pubkey);
+                if should_flush {
+                    written_accounts.insert(*pubkey);
+                } else {
+                    *account_bytes_saved += account.data.len();
+                    *num_accounts_saved += 1;
+                }
+                should_flush
```

Comment:
how about this? HashSet is internally implemented as HashMap<K, ()>; so there should be no downside.

https://github.com/ryoqun/solana/commit/18a0121b7e0fcd2647ed36890aa8401b50bb9639

This is a bit sad, though.

Reply:
argh, wrong comment fixed https://github.com/solana-labs/solana/commit/a2277f575c0eeb6592df19f3a21da7ed4383e08f

Reply:
fyi, I'm encouraged by this comment: https://github.com/rust-lang/rust/issues/60896#issuecomment-721414008